tools/clang/unittests/HLSLTestLib/FileCheckerTest.cpp

```diff
@@ -1286,3 +1301,61 @@ void ParseCommandPartsFromFile(LPCWSTR fileName,
   std::string commands(GetFirstLine(fileName));
   ParseCommandParts(commands.c_str(), fileName, parts);
 }
+
+void ParseFeatures(LPCSTR features, std::vector<TestRequirement> &parts,
+                   bool Requires) {
+  LPCSTR endFeatures = strchr(features, '\0');
+  while (features != endFeatures) {
+    LPCSTR nextStart;
+    LPCSTR thisEnd = strchr(features, ',');
+    if (!thisEnd) {
+      nextStart = thisEnd = endFeatures;
+    } else {
+      nextStart = thisEnd + 2;
+    }
+    parts.emplace_back(std::string(features, thisEnd), Requires);
+    features = nextStart;
+  }
+}
+
+bool ParseRequires(LPCSTR input, LPCWSTR fileName,
+                   std::vector<TestRequirement> &parts) {
+  LPCSTR feature = strstr(input, "REQUIRES: ");
+  if (feature) {
+    feature += strlen("REQUIRES: ");
+    ParseFeatures(feature, parts, true);
+    return true;
+  }
+
+  feature = strstr(input, "UNSUPPORTED: ");
+  if (feature) {
+    feature += strlen("UNSUPPORTED: ");
+    ParseFeatures(feature, parts, false);
+    return true;
+  }
+  return false;
+}
+
+bool TestRequirement::IsSupported(dxc::DxcDllSupport &Support) const {
+  if (Feature == "system-windows") {
+#if _WIN32
+    return Required;
+#else
+    return !Required;
+#endif
+  }
+
+  if (Feature == "HLSL") {
+    return Required;
+  }
+
+  auto versionComponents = strtok(Feature, "-");
+  if (versionComponents.front() == "dxilver") {
+    unsigned RequiredDxilMajor = versionComponents[1][0] - '0';
+    unsigned RequiredDxilMinor = versionComponents[1][2] - '0';
+    auto Result = CheckDxilVer(Support, RequiredDxilMajor, RequiredDxilMinor);
+    return Required ? !Result.AbortPipeline : Result.AbortPipeline;
```

Comment:
I feel like there's a bug in here if this were to be used with `UNSUPPORTED:` due to the fact that the dxil version check is more complicated and contains two checks under the hood.

This checks the Dxil version from the compiler, and the Validator version from the validator.  If the required version is higher than either of these, the check returns true for AbortPipeline.

If the `UNSUPPORTED:` case was needed (don't run this test on newer versions), for whatever reason, then one of these versions could be newer than we want, while the minimum of the two is still below the bar, leading us to run the test with one of the newer component versions that we did not want to run against.

I think the clean solution is to refactor CheckDxilVer contents into a get+compare version function with a flag to switch between compiler/validator (or argument for the CLSID), using that in CheckDxilVer.

For this code, change this code to use that function directly, but two options for the check:
1. check both and ensure that neither of the modules support the given version in the `UNSUPPORTED:` case.
2. add a separate check for validator version, like: `valver-1-6`, so the test can decide (which means updating all the existing cases where you have dxilver to add a corresponding valver with the same version).

I think I prefer option 2, but it's a little more work updating the tests again.